File: packages/raisin-client-py/raisindb/tags.py
"""Tag management operations."""

from typing import TYPE_CHECKING, Any, Dict, List, Optional

from .protocol import RequestType, RequestContext

if TYPE_CHECKING:
    from .database import Database


class Tags:
    """Tag management operations."""

    def __init__(self, database: "Database"):
        """
        Initialize Tags manager.

        Args:
            database: Parent Database instance
        """
        self.database = database

    async def create(
        self, name: str, revision: str, message: Optional[str] = None
    ) -> Any:
        """
        Create a new tag.

        Args:
            name: Tag name
            revision: Revision to tag
            message: Optional tag message

        Returns:
            Created tag
        """
        context = self.database.get_context()
        payload: Dict[str, Any] = {"name": name, "revision": revision}
        if message is not None:
            payload["message"] = message
        return await self.database.client.request_tracker.send_request(
            RequestType.TAG_CREATE, context, payload
        )

    async def get(self, name: str) -> Any:
        """
        Get a tag by name.

        Args:
            name: Tag name

        Returns:
            Tag information
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.TAG_GET, context, payload
        )

    async def list(self) -> List[Any]:
        """
        List all tags.

        Returns:
            List of tags
        """
        context = self.database.get_context()
        result = await self.database.client.request_tracker.send_request(
            RequestType.TAG_LIST, context, {}
        )
        return result if isinstance(result, list) else []

    async def delete(self, name: str) -> Any:
        """
        Delete a tag.

        Args:
            name: Tag name

        Returns:
            Deletion result
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.TAG_DELETE, context, payload
        )

File: packages/raisin-client-js/src/tags.ts
/**
 * Tag management operations
 */

import { RequestContext, RequestType } from './protocol';

export class Tags {
  private sendRequest: (payload: unknown, requestType: string) => Promise<unknown>;

  constructor(
    _context: RequestContext,
    sendRequest: (payload: unknown, requestType: string) => Promise<unknown>
  ) {
    this.sendRequest = sendRequest;
  }

  /**
   * Create a new tag
   */
  async create(name: string, revision: string, message?: string): Promise<unknown> {
    return this.sendRequest(
      {
        name,
        revision,
        message
      },
      RequestType.TagCreate
    );
  }

  /**
   * Get a tag by name
   */
  async get(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.TagGet
    );
  }

  /**
   * List all tags
   */
  async list(): Promise<unknown[]> {
    const result = await this.sendRequest(
      {},
      RequestType.TagList
    );
    return Array.isArray(result) ? result : [];
  }

  /**
   * Delete a tag
   */
  async delete(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.TagDelete
    );
  }
}
